RealmSwift/SectionedResults.swift

```diff
@@ -0,0 +1,1050 @@
+    init(rlmSectionedResult: Collection)
+}
+
+extension RealmSectionedResultImpl {
+    /// The position of the first element in a non-empty collection.
+    /// Identical to endIndex in an empty collection.
+    public var startIndex: Int { 0 }
+
+    /// The collection's "past the end" position.
+    /// endIndex is not a valid argument to subscript, and is always reachable from startIndex by
+    /// zero or more applications of successor().
+    public var endIndex: Int { count }
+    /// The Realm which manages the object.
+    public var realm: Realm? {
+        collection.realm.map(Realm.init)
+    }
+    /// Indicates if the collection can no longer be accessed.
+    public var isInvalidated: Bool {
+        collection.isInvalidated
+    }
+    /// Returns true if this collection is frozen
+    public var isFrozen: Bool {
+        collection.isFrozen
+    }
+    /**
+     Returns a frozen (immutable) snapshot of this collection.
+
+     The frozen copy is an immutable collection which contains the same data as this collection
+    currently contains, but will not update when writes are made to the containing Realm. Unlike
+    live collections, frozen collections can be accessed from any thread.
+
+     - warning: This method cannot be called during a write transaction, or when the containing
+    Realm is read-only.
+     - warning: Holding onto a frozen collection for an extended period while performing write
+     transaction on the Realm may result in the Realm file growing to large sizes. See
+     `Realm.Configuration.maximumNumberOfActiveVersions` for more information.
+    */
+    public func freeze() -> Self {
+        return Self(rlmSectionedResult: collection.freeze())
+    }
+    /// :nodoc:
+    public func thaw() -> Self? {
+        return Self(rlmSectionedResult: collection.thaw())
+    }
+    /// :nodoc:
+    internal typealias ObjcSectionedResultsChange = (RLMSectionedResult?, RLMSectionedResultsChange?, Error?) -> Void
+    /// :nodoc:
+    internal func wrapObserveBlock(_ block: @escaping (RealmSectionedResultsChange<Self>) -> Void) -> ObjcSectionedResultsChange {
+        var col: Self?
+        return { collection, change, error in
+            if col == nil, let collection = collection {
+                col = self.collection === collection ? self : Self(rlmSectionedResult: collection as! Self.Collection)
+            }
+            block(RealmSectionedResultsChange.fromObjc(value: col, change: change, error: error))
+        }
+    }
+}
+
+/// `SectionedResults` is a type safe collection which holds individual `ResultsSection`s as its elements.
+/// The container is lazily evaluated, meaning that if the underlying collection has changed a full recalculation of the section keys will take place.
+/// A `SectionedResults` instance can be observed and it also conforms to `ThreadConfined`.
+public struct SectionedResults<Key: _Persistable & Hashable, T: RealmCollectionValue>: RealmSectionedResultImpl {
```

Comment:
Both of these types need a `Key` typealias so that the generic parameter can be accessed from outside the type. We get `Element` for free from the base collection protocol.